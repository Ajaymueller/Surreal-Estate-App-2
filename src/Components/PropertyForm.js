import React from 'react';
import '../styles/AddProperty.css';

const PropertyForm = ({ handleAddProperty, handleFieldChange, title,
    city,
    type,
    bedrooms,
    bathrooms,
    price,
    email, } ) => {
    return (
        <div>
        <form onSubmit={handleAddProperty} 
        data-testid="property-form">
    <div className="add-property-field">
      <label htmlFor="title">
        Title
        <input
          id="title"
          name="title"
          type="text"
          placeholder="Property tagline"
          value={title}
          onChange={handleFieldChange}
          data-testid="title"

        />
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="city">
        City
        <select
          id="city"
          name="city"
          value={city}
          onChange={handleFieldChange}
          data-testid="city"
        >
          <option value="Manchester">Manchester</option>
          <option value="Leeds">Leeds</option>
          <option value="Sheffield">Sheffield</option>
          <option value="Liverpool">Liverpool</option>
        </select>
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="type">
        Type
        <select
          id="type"
          name="type"
          value={type}
          onChange={handleFieldChange}
          data-testid="type"
        >
          <option value="Flat">Flat</option>
          <option value="Detached">Detached</option>
          <option value="Semi-Detatched">Semi-Detatched</option>
          <option value="Terraced">Terraced</option>
          <option value="End of Terrace">End of Terrace</option>
          <option value="Cottage">Cottage</option>
          <option value="Bungalow">Bungalow</option>
        </select>
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="price">
        Price
        <input
          id="price"
          name="price"
          type="number"
          value={price}
          onChange={handleFieldChange}
          data-testid="price"
        />
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="bedrooms">
        Bedrooms
        <input
          id="bedrooms"
          name="bedrooms"
          type="number"
          value={bedrooms}
          onChange={handleFieldChange}
          data-testid="bedrooms"
        />
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="bathrooms">
        Bathrooms
        <input
          id="bathrooms"
          name="bathrooms"
          type="number"
          value={bathrooms}
          onChange={handleFieldChange}
          data-testid="bathrooms"
        />
      </label>
    </div>
    <div className="add-property-field">
      <label htmlFor="email">
        Email
        <input
          id="email"
          name="email"
          placeholder="Contact email"
          type="email"
          value={email}
          onChange={handleFieldChange}
          data-testid="email"
        />
      </label>
    </div>
    <button type="submit">Add</button>
  </form>
    </div>
    )
}

export default PropertyForm;