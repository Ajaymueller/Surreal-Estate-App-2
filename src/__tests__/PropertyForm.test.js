import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PropertyForm from '../Components/PropertyForm';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';

const props = {
    handleAddProperty: jest.fn(),
    handleFieldChange: jest.fn(),
    title: "randomTitle", 
    city: "Liverpool", 
    type: "Semi-Detached",
    bedrooms: 3,
    bathrooms: 3, 
    price: 250000,
    email: "random@gmail.com",
};

describe("PropertyForm", () => {
    it("component renders with correct props", () => {
        const { asFragment } = render 
        (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        expect(asFragment).toMatchSnapshot();
    });
    it("renders a form", () => {
        const { getByTestId } = 
        render (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        const form = getByTestId("property-form");
        expect(form).toBeInTheDocument();
    });
    it("form should have an add button", () => {
        const { getByRole } = 
        render (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        const btn = getByRole("button");
        expect(btn).toHaveTextContent("Add");
    })
    it("renders form fields correctly", () => {
        const { getByTestId, getByRole } = 
        render (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        expect(getByTestId("title")).toBeInTheDocument()
        expect(getByTestId("city")).toBeInTheDocument()
        expect(getByTestId("type")).toBeInTheDocument()
        expect(getByTestId("bathrooms")).toBeInTheDocument()
        expect(getByTestId("bedrooms")).toBeInTheDocument()
        expect(getByTestId("price")).toBeInTheDocument()
        expect(getByTestId("email")).toBeInTheDocument()
        expect(getByRole('button')).toBeInTheDocument()
    })
    it("should update the state when fields changed", () => {
        const { getByTestId } = render 
        (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        const handleFieldChange = props.handleFieldChange;
        const title = getByTestId("title");
        const city = getByTestId("city");
        const type = getByTestId("type");
        const bathrooms = getByTestId("bathrooms");
        const bedrooms = getByTestId("bedrooms");
        const price = getByTestId("price");
        const email = getByTestId("email");
        fireEvent.change(title, { target: { value: "changedValue" } });
        fireEvent.change(city, { target: { value: "changedValue" } });
        fireEvent.change(type, { target: { value: "changedValue" } });
        fireEvent.change(bathrooms, { target: { value: "changedValue" } });
        fireEvent.change(bedrooms, { target: { value: "changedValue" } });
        fireEvent.change(price, { target: { value: "changedValue" } });
        fireEvent.change(email, { target: { value: "changedValue" } });
        expect(handleFieldChange).toHaveBeenCalled();
    });
    it("should call handleAddProperty on form submission", () => {
        const { getByTestId } = render 
        (<MemoryRouter><PropertyForm {...props} /></MemoryRouter> )
        const handleAddProperty = props.handleAddProperty;
        fireEvent.submit(getByTestId("property-form"));
        expect(handleAddProperty).toHaveBeenCalled();
    });
});