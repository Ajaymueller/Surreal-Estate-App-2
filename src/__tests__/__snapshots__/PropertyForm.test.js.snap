exports[`PropertyForm component renders with correct props 1`] = `[Function]`;
